### PR TITLE
build: Remove bundled protoc and build from source

### LINKED ISSUE #575: Supply chain security with embedded protoc binaries
We're thinking about using prost over at @MaterializeInc, but the supply chain security of the embedded `protoc` binaries has me very nervous.

Specifically: what sort of validation are you doing before you merge PRs to check that the binaries supplied by third parties are identical to the binaries published by google/protobuf? I did some spot checks and they look identical, but I didn't see anything in your CI pipeline that would assert this. 

We'd honestly probably want to opt out of the bundled protoc deps entirely. I know that's already possible if you specify `PROTOC`, but I'd feel better if there were a way to not even bring the binaries into our dependency tree. Otherwise it seems too easy for accidents to happen; e.g., if that env var somehow gets unset. 

How would y'all feel about adopting an approach similar to rust-protobuf, where the bundled binaries are kept in a separate crate (e.g., `prost-build-vendored`)? Using that crate could be enabled by default, but we could keep it off @MaterializeInc and lint for its absence using `cargo deny`, and I'd sleep a good bit easier at night.

I've also been toying with the idea of building something like `protobuf-src` for Rust (a la `openssl-src`) that builds libprotobuf from source and makes its protoc available to downstream libraries. That could be a nice alternative for `prost-build` to expose.

### PR DESCRIPTION
**Pull request by LucioFranco**
This is an initial attempt to remove bundled binaries from `prost-build` and to instead compile protobuf from source. This is a much safer approach and reduces prost's vulnerability to supply chain attacks.

Reference https://github.com/tokio-rs/prost/issues/575#issuecomment-1078153519 for more information on how this is implemented.

This also adds a `test-vendored` crate which ensures that vendoring works.

Closes #575

### CODE REVIEW

File: prost-build/Cargo.toml [hunk 1]
```diff
@@ -27,6 +31,9 @@ regex = { version = "1.5.5", default-features = false, features = ["std", "unico
 
 [build-dependencies]
 which = { version = "4", default-features = false }
+cfg-if = "1"
+autotools = "0.2"
```
Comment (LucioFranco) [inline prost-build/Cargo.toml:35]: needs to be removed

File: test-vendored/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "test-vendored"
+version = "0.1.0"
+edition = "2021"
```
Comment (LucioFranco) [inline test-vendored/Cargo.toml:4]: should get a publish = false

File: prost-build/third-party/update-vendored-protobuf.sh [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+mv "$TEMPDIR/linux-x86_64/include" "$DIR/include/"
+
+# # Update the Protocol Buffers license.
+# mkdir "$TEMPDIR/src"
+# curl --proto '=https' --tlsv1.2 -sSfL \
+#   "https://github.com/protocolbuffers/protobuf/archive/v$VERSION.zip" \
+#   -o "$TEMPDIR/src/protobuf.zip"
+# unzip "$TEMPDIR/src/protobuf.zip" -d "$TEMPDIR/src"
+
+# mv "$TEMPDIR/src/protobuf-$VERSION" "$DIR/protobuf-src"
```
Comment (LucioFranco) [inline prost-build/third-party/update-vendored-protobuf.sh:34]: Remove this since we track from git

**Comment by Thomasdezeeuw** [comment]
I just want to note that this change (I think) requires `cmake` to be install to build prost, which broke our build pipeline.

**Pull request merged**